File: Chapter - 2 -  Linked List/6. Palindrome/Palindrome.cpp
#include <iostream>
using namespace std;
	
struct Node{
	int data;
	Node* next;
	Node(int x){
		data = x;
		next = NULL;
	}
};
	
Node* pushToLast(Node* head, int data){
	Node* node = new Node(data);
	Node* temp = head;
	if(head == NULL){
		head = node;
	}else{
		while(temp && temp->next) temp = temp->next;
		temp->next = node;
	}
	return head;
}
	
Node* pushToFront(Node* head, int data){
	Node* node = new Node(data);
	Node* temp = head;
	if(temp == NULL) head = node;
	else{
		node->next = head;
		head = node;
	}
	return head;
}
	
int getLength(Node* head){
	Node* temp = head;
	int count = 0;
	while(temp){
		count++;
		temp = temp->next;
	}
	return count;
}
	
void printList(Node* head){
	Node* temp = head;
	while(temp){
		cout << temp->data << " ";
		temp = temp->next;
	}
}

void palindrome(Node *head){
	Node* ans = NULL;
	Node* temp = head;
	while(temp){
		ans = pushToFront(ans, temp->data);
		temp = temp->next;
	}
	temp = head;
	while(temp && ans){
		if(temp->data != ans->data){
			cout << "Not a Plaindrome";
			return;
		}
		temp = temp->next;
		ans = ans->next;
	}
	cout << "Palindrome !!!";
}

int main() {
	int n;
	cin >> n;
	Node* head = NULL;
	for(int i = 0 ; i < n ; i++){
		int x;
		cin >> x;
		head = pushToLast(head, x);
	}
	
	palindrome(head);
	
	return 0;
}

File: Chapter - 2 -  Linked List/6. Palindrome/Palindrome.java
class Palindrome {
    static ListNode head = null;

    public static ListNode reverse(ListNode head) {
        if (head == null)
            return head;

        ListNode current = head;
        ListNode temp;
        while (current.next != null) {
            temp = current.next.next;
            current.next.next = head;
            head = current.next;
            current.next = temp;
        }
        return head;
    }

    public static void print(ListNode head) {
        ListNode current = head;
        while (current != null) {
            System.out.print(current.val + " ");
            current = current.next;
        }
        System.out.println();
    }

    public static void add(int data) {
        if (head == null) {
            head = new ListNode(data);
            return;
        }
        ListNode curr = head;
        while (curr.next != null) {
            curr = curr.next;
        }
        curr.next = new ListNode(data);
    }

    public static void main(String[] args) {
        add(3);
        add(5);
        add(31);
        add(90);
        add(43);

        print(head);
        head = reverse(head);
        print(head);
    }

    static class ListNode {
        int val;
        ListNode next;

        ListNode(int x) {
            val = x;
        }
    }
}